src/librustc_middle/ty/sty.rs

```diff
@@ -2333,14 +2333,41 @@ impl<'tcx> Const<'tcx> {
         assert_eq!(self.ty, ty);
         let size = tcx.layout_of(param_env.with_reveal_all().and(ty)).ok()?.size;
         // if `ty` does not depend on generic parameters, use an empty param_env
-        self.eval(tcx, param_env).val.try_to_bits(size)
+        self.val.eval(tcx, param_env).try_to_bits(size)
     }
 
     #[inline]
     /// Tries to evaluate the constant if it is `Unevaluated`. If that doesn't succeed, return the
     /// unevaluated constant.
     pub fn eval(&self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>) -> &Const<'tcx> {
-        if let ConstKind::Unevaluated(did, substs, promoted) = self.val {
+        if let Some(val) = self.val.try_eval(tcx, param_env) {
+            match val {
+                Ok(val) => Const::from_value(tcx, val, self.ty),
+                Err(ErrorReported) => tcx.const_error(self.ty),
+            }
+        } else {
+            self
+        }
+    }
+}
+
+impl<'tcx> ConstKind<'tcx> {
```

Comment:
So we now have an `impl ConstKind` block between two `impl Const` block and *above* `enum ConstKind`. Could you clean that up a bit?

Reply:
The reason I did this was to make the diff small. But as a separate commit just for moving it that would work I guess.

Reply:
The small diff makes it more confusing -- as you said yourself, this is not the same function.

Reply:
Not sure if you had in mind what I did (I went a bit overboard), but at least everything has its own place now.

Reply:
I mean now the diff is indeed unreadable, but splitting that big file is probably a good idea.^^

Reply:
I kept the commits clean, so if you look commit-by-commit, everything stays readable except for the split commit, which is entirely without functional changes.